aten/src/ATen/native/cuda/Embedding.cu

```diff
@@ -0,0 +1,348 @@
+  int64_t* count, int64_t numel, int64_t stride, int padding_idx) {
+
+  using accscalar_t = cuda::acc_type<scalar_t>;
+  int idx = blockIdx.x * 4 + threadIdx.y;
+
+  // Each warp is responsible for an input into the LookupTable.
+  // If the preceding input has the same as this input, then the warp
+  // exits immediately. The warp also processes subsequent inputs with the
+  // same value.
+  //
+  // Input Warp
+  // 1     <warp 1>
+  // 1     <warp 1> (<warp 2> exits without doing any work)
+  // 5     <warp 3>
+  // 8     <warp 4>
+
+  // Number of values proceessed by each thread (grain size)
+  const int SZ = 4;
+
+  if (idx < numel
+      && (idx == 0 || input[idx] != input[idx - 1])
+      && input[idx] != padding_idx) {
+    do {
+      const int start_feature = threadIdx.x + blockIdx.y * blockDim.x * SZ;
+      const int weight_row = ((int) input[idx]) * stride;
+      const int grad_row = ((int) indices[idx]) * stride;
+      const accscalar_t scale = count ? (accscalar_t)1.0 / count[idx] : 1.0;
+
+      accscalar_t gradient[SZ];
+      accscalar_t weight[SZ];
+
+      #pragma unroll
+      for (int ii = 0; ii < SZ; ii++) {
+        int feature_dim = start_feature + ii * WARP_SIZE;
+        if (feature_dim < stride) {
+          gradient[ii] = scalar_cast<accscalar_t>(grad_output[grad_row + feature_dim]);
+          weight[ii] = scalar_cast<accscalar_t>(grad_weight[weight_row + feature_dim]);
+        }
+      }
+
+      #pragma unroll
+      for (int ii = 0; ii < SZ; ii++) {
+        weight[ii] += gradient[ii] * scale;
+      }
+
+      #pragma unroll
+      for (int ii = 0; ii < SZ; ii++) {
+        int feature_dim = start_feature + ii * WARP_SIZE;
+        if (feature_dim < stride) {
+            grad_weight[weight_row + feature_dim] = scalar_cast<scalar_t>(weight[ii]);
+        }
+      }
+
+      idx++;
+    } while (idx < numel && input[idx] == input[idx - 1]);
+  }
+}
+
+/* Calculate norms of the rows of weight_ptr given by idx_ptr and capture them in norms */
+template <typename scalar_t, typename accscalar_t>
+__global__ void renorm_kernel(
+    scalar_t* weights, int64_t* indices, accscalar_t max_norm,
+    accscalar_t norm_type, int dim) {
+
+  // Some casting hacks since dynamic shared memory and templates don't work together:
+  extern __shared__ unsigned char smem[];
+  auto sdata = reinterpret_cast<accscalar_t*>(smem);
+
+  int tid = threadIdx.x;
+  int base_index = indices[blockIdx.x] * dim;
+
+  accscalar_t v = 0;
+  for (int i = tid; i < dim; i += blockDim.x) {
+    auto x = scalar_cast<accscalar_t>(weights[base_index + i]);
+    if (norm_type == 1) {
+      v += std::abs(x);
+    } else if (norm_type == 2) {
+      v += x * x;
+    } else {
+      v += std::pow(x, norm_type);
+    }
+  }
+
+  using Op = ReduceAdd<accscalar_t, accscalar_t>;
+  v = reduceBlock<accscalar_t>(sdata, blockDim.x, v, Op(), 0);
+
+  if (tid == 0) {
+    sdata[0] = std::pow(v, 1.0 / scalar_cast<accscalar_t>(norm_type));
```

Comment:
MSVC seems to complain about the following:
```
error: calling a __host__ function("pow<float, double, void> ") from a __global__ function("at::native::_NV_ANON_NAMESPACE::renorm_kernel<float, float> ") is not allowed
error: identifier "pow<float, double, void> " is undefined in device code
error: calling a __host__ function("pow<float, double, void> ") from a __global__ function("at::native::_NV_ANON_NAMESPACE::renorm_kernel< ::__half, float> ") is not allowed
error: identifier "pow<float, double, void> " is undefined in device code
```

I tried to add `#include "math.h"` at the top of file to include the CUDA math library, but it seems that calling `pow(...)` inside the `at::native` namespace will be calling the ATen tensor `pow(...)` instead.